File: .net/vb.net/Server/ServerDataProcessor.vb
Friend Class ServerDataProcessor
    Inherits wclCustomServerClientDataProcessor

    Private FDecoder As CommandDecoder

#Region "Data decoder events."
    Private Sub GetArray(Sender As Object)
        If OnGetArrayEvent IsNot Nothing Then
            RaiseEvent OnGetArray(Me)
        End If
    End Sub

    Private Sub GetInt64(Sender As Object)
        If OnGetInt64Event IsNot Nothing Then
            RaiseEvent OnGetInt64(Me)
        End If
    End Sub

    Private Sub GetInt32(Sender As Object)
        If OnGetInt32Event IsNot Nothing Then
            RaiseEvent OnGetInt32(Me)
        End If
    End Sub

    Private Sub GetInt16(Sender As Object)
        If OnGetInt16Event IsNot Nothing Then
            RaiseEvent OnGetInt16(Me)
        End If
    End Sub

    Private Sub GetSByte(Sender As Object)
        If OnGetSByteEvent IsNot Nothing Then
            RaiseEvent OnGetSByte(Me)
        End If
    End Sub

    Private Sub GetUInt64(Sender As Object)
        If OnGetUInt64Event IsNot Nothing Then
            RaiseEvent OnGetUInt64(Me)
        End If
    End Sub

    Private Sub GetUInt32(Sender As Object)
        If OnGetUInt32Event IsNot Nothing Then
            RaiseEvent OnGetUInt32(Me)
        End If
    End Sub

    Private Sub GetUInt16(Sender As Object)
        If OnGetUInt16Event IsNot Nothing Then
            RaiseEvent OnGetUInt16(Me)
        End If
    End Sub

    Private Sub GetByte(Sender As Object)
        If OnGetByteEvent IsNot Nothing Then
            RaiseEvent OnGetByte(Me)
        End If
    End Sub

    Private Sub GetString(Sender As Object)
        If OnGetStringEvent IsNot Nothing Then
            RaiseEvent OnGetString(Me)
        End If
    End Sub

    Private Sub ArrayReceived(Sender As Object, Data As Byte())
        If OnArrayReceivedEvent IsNot Nothing Then
            RaiseEvent OnArrayReceived(Me, Data)
        End If
    End Sub

    Private Sub Int64Received(Sender As Object, Data As Int64)
        If OnInt64ReceivedEvent IsNot Nothing Then
            RaiseEvent OnInt64Received(Me, Data)
        End If
    End Sub

    Private Sub Int32Received(Sender As Object, Data As Int32)
        If OnInt32ReceivedEvent IsNot Nothing Then
            RaiseEvent OnInt32Received(Me, Data)
        End If
    End Sub

    Private Sub Int16Received(Sender As Object, Data As Int16)
        If OnInt16ReceivedEvent IsNot Nothing Then
            RaiseEvent OnInt16Received(Me, Data)
        End If
    End Sub

    Private Sub SByteReceived(Sender As Object, Data As SByte)
        If OnSByteReceivedEvent IsNot Nothing Then
            RaiseEvent OnSByteReceived(Me, Data)
        End If
    End Sub

    Private Sub UInt64Received(Sender As Object, Data As UInt64)
        If OnUInt64ReceivedEvent IsNot Nothing Then
            RaiseEvent OnUInt64Received(Me, Data)
        End If
    End Sub

    Private Sub UInt32Received(Sender As Object, Data As UInt32)
        If OnUInt32ReceivedEvent IsNot Nothing Then
            RaiseEvent OnUInt32Received(Me, Data)
        End If
    End Sub

    Private Sub UInt16Received(Sender As Object, Data As UInt16)
        If OnUInt16ReceivedEvent IsNot Nothing Then
            RaiseEvent OnUInt16Received(Me, Data)
        End If
    End Sub

    Private Sub ByteReceived(Sender As Object, Data As Byte)
        If OnByteReceivedEvent IsNot Nothing Then
            RaiseEvent OnByteReceived(Me, Data)
        End If
    End Sub

    Private Sub StringReceived(Sender As Object, Data As String)
        If OnStringReceivedEvent IsNot Nothing Then
            RaiseEvent OnStringReceived(Me, Data)
        End If
    End Sub
#End Region

    Protected Overrides Sub ProcessData(Data As Byte())
        FDecoder.ProcessData(Data)
    End Sub

    Public Sub New(Connection As wclServerClientDataConnection)
        MyBase.New(Connection)

        FDecoder = New CommandDecoder()

        AddHandler FDecoder.OnByteReceived, AddressOf ByteReceived
        AddHandler FDecoder.OnUInt16Received, AddressOf UInt16Received
        AddHandler FDecoder.OnUInt32Received, AddressOf UInt32Received
        AddHandler FDecoder.OnUInt64Received, AddressOf UInt64Received
        AddHandler FDecoder.OnSByteReceived, AddressOf SByteReceived
        AddHandler FDecoder.OnInt16Received, AddressOf Int16Received
        AddHandler FDecoder.OnInt32Received, AddressOf Int32Received
        AddHandler FDecoder.OnInt64Received, AddressOf Int64Received
        AddHandler FDecoder.OnArrayReceived, AddressOf ArrayReceived
        AddHandler FDecoder.OnStringReceived, AddressOf StringReceived

        AddHandler FDecoder.OnGetByte, AddressOf GetByte
        AddHandler FDecoder.OnGetUInt16, AddressOf GetUInt16
        AddHandler FDecoder.OnGetUInt32, AddressOf GetUInt32
        AddHandler FDecoder.OnGetUInt64, AddressOf GetUInt64
        AddHandler FDecoder.OnGetSByte, AddressOf GetSByte
        AddHandler FDecoder.OnGetInt16, AddressOf GetInt16
        AddHandler FDecoder.OnGetInt32, AddressOf GetInt32
        AddHandler FDecoder.OnGetInt64, AddressOf GetInt64
        AddHandler FDecoder.OnGetArray, AddressOf GetArray
        AddHandler FDecoder.OnGetString, AddressOf GetString

        OnByteReceivedEvent = Nothing
        OnUInt16ReceivedEvent = Nothing
        OnUInt32ReceivedEvent = Nothing
        OnUInt64ReceivedEvent = Nothing
        OnSByteReceivedEvent = Nothing
        OnInt16ReceivedEvent = Nothing
        OnInt32ReceivedEvent = Nothing
        OnInt64ReceivedEvent = Nothing
        OnArrayReceivedEvent = Nothing
        OnStringReceivedEvent = Nothing

        OnGetByteEvent = Nothing
        OnGetUInt16Event = Nothing
        OnGetUInt32Event = Nothing
        OnGetUInt64Event = Nothing
        OnGetSByteEvent = Nothing
        OnGetInt16Event = Nothing
        OnGetInt32Event = Nothing
        OnGetInt64Event = Nothing
        OnGetArrayEvent = Nothing
        OnGetStringEvent = Nothing
    End Sub

#Region "Write data"
    Public Function WriteByte(Data As Byte) As Int32
        Return Write(CommandBuilder.Create(Data))
    End Function

    Public Function WriteSByte(Data As SByte) As Int32
        Return Write(CommandBuilder.Create(Data))
    End Function

    Public Function WriteUInt16(Data As UInt16) As Int32
        Return Write(CommandBuilder.Create(Data))
    End Function

    Public Function WriteInt16(Data As Int16) As Int32
        Return Write(CommandBuilder.Create(Data))
    End Function

    Public Function WriteUInt32(Data As UInt32) As Int32
        Return Write(CommandBuilder.Create(Data))
    End Function

    Public Function WriteInt32(Data As Int32) As Int32
        Return Write(CommandBuilder.Create(Data))
    End Function

    Public Function WriteUInt64(Data As UInt64) As Int32
        Return Write(CommandBuilder.Create(Data))
    End Function

    Public Function WriteInt64(Data As Int64) As Int32
        Return Write(CommandBuilder.Create(Data))
    End Function

    Public Function WriteArray(Data As Byte()) As Int32
        If Data Is Nothing Or Data.Length = 0 Or CUInt(Data.Length) > UInt16.MaxValue - 3 Then
            Return wclErrors.WCL_E_INVALID_ARGUMENT
        End If

        Return Write(CommandBuilder.Create(Data))
    End Function

    Public Function WriteString(Data As String) As Int32
        If Data Is Nothing Or Data.Length = 0 Or CUInt(Data.Length) > UInt16.MaxValue - 3 Then
            Return wclErrors.WCL_E_INVALID_ARGUMENT
        End If

        Return Write(CommandBuilder.Create(Data))
    End Function

    Public Function WriteError([Error] As Int32) As Int32
        Return Write(CommandBuilder.CreateError([Error]))
    End Function
#End Region

#Region "Events."
#Region "Received events."
    Public Event OnByteReceived As ByteReceived
    Public Event OnUInt16Received As UInt16Received
    Public Event OnUInt32Received As UInt32Received
    Public Event OnUInt64Received As UInt64Received
    Public Event OnSByteReceived As SByteReceived
    Public Event OnInt16Received As Int16Received
    Public Event OnInt32Received As Int32Received
    Public Event OnInt64Received As Int64Received
    Public Event OnArrayReceived As ArrayReceived
    Public Event OnStringReceived As StringReceived
#End Region

#Region "Get events."
    Public Event OnGetByte As DataEvent
    Public Event OnGetUInt16 As DataEvent
    Public Event OnGetUInt32 As DataEvent
    Public Event OnGetUInt64 As DataEvent
    Public Event OnGetSByte As DataEvent
    Public Event OnGetInt16 As DataEvent
    Public Event OnGetInt32 As DataEvent
    Public Event OnGetInt64 As DataEvent
    Public Event OnGetArray As DataEvent
    Public Event OnGetString As DataEvent
#End Region
#End Region
End Class

File: .net/c#/Server/ServerDataProcessor.cs
using System;

using wclCommon;
using wclCommunication;

using RfCommClientServer;

namespace RfCommServer
{
    internal class ServerDataProcessor : wclCustomServerClientDataProcessor
    {
        private CommandDecoder FDecoder;

        #region Data decoder events.
        private void GetArray(Object Sender)
        {
            if (OnGetArray != null)
                OnGetArray(this);
        }

        private void GetInt64(Object Sender)
        {
            if (OnGetInt64 != null)
                OnGetInt64(this);
        }

        private void GetInt32(Object Sender)
        {
            if (OnGetInt32 != null)
                OnGetInt32(this);
        }

        private void GetInt16(Object Sender)
        {
            if (OnGetInt16 != null)
                OnGetInt16(this);
        }

        private void GetSByte(Object Sender)
        {
            if (OnGetSByte != null)
                OnGetSByte(this);
        }

        private void GetUInt64(Object Sender)
        {
            if (OnGetUInt64 != null)
                OnGetUInt64(this);
        }

        private void GetUInt32(Object Sender)
        {
            if (OnGetUInt32 != null)
                OnGetUInt32(this);
        }

        private void GetUInt16(Object Sender)
        {
            if (OnGetUInt16 != null)
                OnGetUInt16(this);
        }

        private void GetByte(Object Sender)
        {
            if (OnGetByte != null)
                OnGetByte(this);
        }

        private void GetString(Object Sender)
        {
            if (OnGetString != null)
                OnGetString(this);
        }

        private void ArrayReceived(Object Sender, Byte[] Data)
        {
            if (OnArrayReceived != null)
                OnArrayReceived(this, Data);
        }

        private void Int64Received(Object Sender, Int64 Data)
        {
            if (OnInt64Received != null)
                OnInt64Received(this, Data);
        }

        private void Int32Received(Object Sender, Int32 Data)
        {
            if (OnInt32Received != null)
                OnInt32Received(this, Data);
        }

        private void Int16Received(Object Sender, Int16 Data)
        {
            if (OnInt16Received != null)
                OnInt16Received(this, Data);
        }

        private void SByteReceived(Object Sender, SByte Data)
        {
            if (OnSByteReceived != null)
                OnSByteReceived(this, Data);
        }

        private void UInt64Received(Object Sender, UInt64 Data)
        {
            if (OnUInt64Received != null)
                OnUInt64Received(this, Data);
        }

        private void UInt32Received(Object Sender, UInt32 Data)
        {
            if (OnUInt32Received != null)
                OnUInt32Received(this, Data);
        }

        private void UInt16Received(Object Sender, UInt16 Data)
        {
            if (OnUInt16Received != null)
                OnUInt16Received(this, Data);
        }

        private void ByteReceived(Object Sender, Byte Data)
        {
            if (OnByteReceived != null)
                OnByteReceived(this, Data);
        }

        private void StringReceived(Object Sender, String Data)
        {
            if (OnStringReceived != null)
                OnStringReceived(this, Data);
        }
        #endregion

        protected override void ProcessData(Byte[] Data)
        {
            FDecoder.ProcessData(Data);
        }

        public ServerDataProcessor(wclServerClientDataConnection Connection)
            : base(Connection)
        {
            FDecoder = new CommandDecoder();

            FDecoder.OnByteReceived += ByteReceived;
            FDecoder.OnUInt16Received += UInt16Received;
            FDecoder.OnUInt32Received += UInt32Received;
            FDecoder.OnUInt64Received += UInt64Received;
            FDecoder.OnSByteReceived += SByteReceived;
            FDecoder.OnInt16Received += Int16Received;
            FDecoder.OnInt32Received += Int32Received;
            FDecoder.OnInt64Received += Int64Received;
            FDecoder.OnArrayReceived += ArrayReceived;
            FDecoder.OnStringReceived += StringReceived;

            FDecoder.OnGetByte += GetByte;
            FDecoder.OnGetUInt16 += GetUInt16;
            FDecoder.OnGetUInt32 += GetUInt32;
            FDecoder.OnGetUInt64 += GetUInt64;
            FDecoder.OnGetSByte += GetSByte;
            FDecoder.OnGetInt16 += GetInt16;
            FDecoder.OnGetInt32 += GetInt32;
            FDecoder.OnGetInt64 += GetInt64;
            FDecoder.OnGetArray += GetArray;
            FDecoder.OnGetString += GetString;

            OnByteReceived = null;
            OnUInt16Received = null;
            OnUInt32Received = null;
            OnUInt64Received = null;
            OnSByteReceived = null;
            OnInt16Received = null;
            OnInt32Received = null;
            OnInt64Received = null;
            OnArrayReceived = null;
            OnStringReceived = null;

            OnGetByte = null;
            OnGetUInt16 = null;
            OnGetUInt32 = null;
            OnGetUInt64 = null;
            OnGetSByte = null;
            OnGetInt16 = null;
            OnGetInt32 = null;
            OnGetInt64 = null;
            OnGetArray = null;
            OnGetString = null;
        }

        #region Write data
        public Int32 WriteByte(Byte Data)
        {
            return Write(CommandBuilder.Create(Data));
        }

        public Int32 WriteSByte(SByte Data)
        {
            return Write(CommandBuilder.Create(Data));
        }

        public Int32 WriteUInt16(UInt16 Data)
        {
            return Write(CommandBuilder.Create(Data));
        }

        public Int32 WriteInt16(Int16 Data)
        {
            return Write(CommandBuilder.Create(Data));
        }

        public Int32 WriteUInt32(UInt32 Data)
        {
            return Write(CommandBuilder.Create(Data));
        }

        public Int32 WriteInt32(Int32 Data)
        {
            return Write(CommandBuilder.Create(Data));
        }

        public Int32 WriteUInt64(UInt64 Data)
        {
            return Write(CommandBuilder.Create(Data));
        }

        public Int32 WriteInt64(Int64 Data)
        {
            return Write(CommandBuilder.Create(Data));
        }
        
        public Int32 WriteArray(Byte[] Data)
        {
            if (Data == null || Data.Length == 0 || (UInt32)Data.Length > UInt16.MaxValue - 3)
                return wclErrors.WCL_E_INVALID_ARGUMENT;

            return Write(CommandBuilder.Create(Data));
        }

        public Int32 WriteString(String Data)
        {
            if (Data == null || Data.Length == 0 || (UInt32)Data.Length > UInt16.MaxValue - 3)
                return wclErrors.WCL_E_INVALID_ARGUMENT;

            return Write(CommandBuilder.Create(Data));
        }
        
        public Int32 WriteError(Int32 Error)
        {
            return Write(CommandBuilder.CreateError(Error));
        }
        #endregion

        #region Events.
        #region Received events.
        public event ByteReceived OnByteReceived;
        public event UInt16Received OnUInt16Received;
        public event UInt32Received OnUInt32Received;
        public event UInt64Received OnUInt64Received;
        public event SByteReceived OnSByteReceived;
        public event Int16Received OnInt16Received;
        public event Int32Received OnInt32Received;
        public event Int64Received OnInt64Received;
        public event ArrayReceived OnArrayReceived;
        public event StringReceived OnStringReceived;
        #endregion

        #region Get events.
        public event DataEvent OnGetByte;
        public event DataEvent OnGetUInt16;
        public event DataEvent OnGetUInt32;
        public event DataEvent OnGetUInt64;
        public event DataEvent OnGetSByte;
        public event DataEvent OnGetInt16;
        public event DataEvent OnGetInt32;
        public event DataEvent OnGetInt64;
        public event DataEvent OnGetArray;
        public event DataEvent OnGetString;
        #endregion 
        #endregion
    }
}
